clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2529,13 +2419,20 @@ class SyclOptReportCreator : public SyclKernelFieldHandler {
   SyclOptReportCreator(Sema &S, SyclKernelDeclCreator &DC, SourceLocation Loc)
       : SyclKernelFieldHandler(S), DC(DC), KernelInvocationLoc(Loc) {}
 
-  bool handleSyclAccessorType(FieldDecl *FD, QualType FieldTy) final {
+  bool handleSyclSpecialType(FieldDecl *FD, QualType FieldTy) final {
+    const auto *RecordDecl = FieldTy->getAsCXXRecordDecl();
+    if (getMethodByName(RecordDecl, FinalizeMethodName))
+      return handleSpecialType(
+          FD, FieldTy, KernelArgDescription(KernelArgDescription::Stream));
+    if (dyn_cast<ClassTemplateSpecializationDecl>(FieldTy->getAsRecordDecl()))
+      return handleSpecialType(
+          FD, FieldTy, KernelArgDescription(KernelArgDescription::Accessor));
     return handleSpecialType(
-        FD, FieldTy, KernelArgDescription(KernelArgDescription::Accessor));
+        FD, FieldTy, KernelArgDescription(KernelArgDescription::Sampler));
```

Comment:
What is the goal for this separation? NOTE: I don't think it's needed and code doesn't seem to be reliable.

Reply:
@bader Thanks for the review. I will take a look. The opt-report needs to report the class type in the report, and that's done via the KernelArgDescrition. I will take a look at the function more closely, to get rid of the distinction here but not loose the information in the report.

Reply:
@elizabethandrews @Pennycook  Do we need for the opt-report to actually report about the specific type (sampler/stream/accessor) of the special class or can it just report that the class is special? So we would be removing from KernelArgDescription Accessor, Sampler and Stream entries and replacing them with Special? Would that be OK for the report?

Reply:
The main question is: should `handle*` methods do different things for different types or not?
If we need only print the name of the special type to the report, it can be done in a single `handle*` function.

Reply:
I think that should be fine since we also emit the name of the accessor,  but I will confirm and let you know